File: backend/app/schemas/sales.py
"""Pydantic schemas for the Sales / Accounts Receivable system."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


# ---------------------------------------------------------------------------
# Quote Lines
# ---------------------------------------------------------------------------


class QuoteLineCreate(BaseModel):
    product_id: str | None = None
    description: str
    quantity: Decimal
    unit_price: Decimal
    sort_order: int = 0


class QuoteLineResponse(BaseModel):
    id: str
    quote_id: str
    product_id: str | None = None
    product_name: str | None = None
    description: str
    quantity: Decimal
    unit_price: Decimal
    line_total: Decimal
    sort_order: int

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Quotes
# ---------------------------------------------------------------------------


class QuoteCreate(BaseModel):
    customer_id: str
    quote_date: datetime
    expiry_date: datetime
    payment_terms: str | None = None
    tax_rate: Decimal = Decimal("0.00")
    notes: str | None = None
    lines: list[QuoteLineCreate] = []


class QuoteUpdate(BaseModel):
    status: str | None = None
    expiry_date: datetime | None = None
    payment_terms: str | None = None
    notes: str | None = None


class QuoteResponse(BaseModel):
    id: str
    company_id: str
    number: str
    customer_id: str
    customer_name: str | None = None
    status: str
    quote_date: datetime
    expiry_date: datetime
    payment_terms: str | None = None
    subtotal: Decimal
    tax_rate: Decimal
    tax_amount: Decimal
    total: Decimal
    notes: str | None = None
    created_by: str | None = None
    created_by_name: str | None = None
    created_at: datetime
    modified_at: datetime | None = None
    lines: list[QuoteLineResponse] = []

    class Config:
        from_attributes = True


class PaginatedQuotes(BaseModel):
    items: list[QuoteResponse]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# Sales Order Lines
# ---------------------------------------------------------------------------


class SalesOrderLineCreate(BaseModel):
    product_id: str | None = None
    description: str
    quantity: Decimal
    unit_price: Decimal
    sort_order: int = 0


class SalesOrderLineResponse(BaseModel):
    id: str
    sales_order_id: str
    product_id: str | None = None
    product_name: str | None = None
    description: str
    quantity: Decimal
    quantity_shipped: Decimal
    unit_price: Decimal
    line_total: Decimal
    sort_order: int

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Sales Orders
# ---------------------------------------------------------------------------


class SalesOrderCreate(BaseModel):
    customer_id: str
    quote_id: str | None = None
    order_date: datetime
    required_date: datetime | None = None
    payment_terms: str | None = None
    tax_rate: Decimal = Decimal("0.00")
    ship_to_name: str | None = None
    ship_to_address: str | None = None
    notes: str | None = None
    lines: list[SalesOrderLineCreate] = []


class SalesOrderUpdate(BaseModel):
    status: str | None = None
    required_date: datetime | None = None
    shipped_date: datetime | None = None
    notes: str | None = None


class SalesOrderResponse(BaseModel):
    id: str
    company_id: str
    number: str
    customer_id: str
    customer_name: str | None = None
    quote_id: str | None = None
    status: str
    order_date: datetime
    required_date: datetime | None = None
    shipped_date: datetime | None = None
    payment_terms: str | None = None
    subtotal: Decimal
    tax_rate: Decimal
    tax_amount: Decimal
    total: Decimal
    ship_to_name: str | None = None
    ship_to_address: str | None = None
    notes: str | None = None
    created_by: str | None = None
    created_by_name: str | None = None
    created_at: datetime
    modified_at: datetime | None = None
    lines: list[SalesOrderLineResponse] = []

    class Config:
        from_attributes = True


class PaginatedSalesOrders(BaseModel):
    items: list[SalesOrderResponse]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# Invoice Lines
# ---------------------------------------------------------------------------


class InvoiceLineCreate(BaseModel):
    product_id: str | None = None
    description: str
    quantity: Decimal
    unit_price: Decimal
    sort_order: int = 0


class InvoiceLineResponse(BaseModel):
    id: str
    invoice_id: str
    product_id: str | None = None
    product_name: str | None = None
    description: str
    quantity: Decimal
    unit_price: Decimal
    line_total: Decimal
    sort_order: int

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Invoices
# ---------------------------------------------------------------------------


class InvoiceCreate(BaseModel):
    customer_id: str
    sales_order_id: str | None = None
    invoice_date: datetime
    due_date: datetime
    payment_terms: str | None = None
    tax_rate: Decimal = Decimal("0.00")
    notes: str | None = None
    lines: list[InvoiceLineCreate] = []


class InvoiceUpdate(BaseModel):
    status: str | None = None
    notes: str | None = None


class InvoiceResponse(BaseModel):
    id: str
    company_id: str
    number: str
    customer_id: str
    customer_name: str | None = None
    sales_order_id: str | None = None
    status: str
    invoice_date: datetime
    due_date: datetime
    payment_terms: str | None = None
    subtotal: Decimal
    tax_rate: Decimal
    tax_amount: Decimal
    total: Decimal
    amount_paid: Decimal
    balance_remaining: Decimal
    notes: str | None = None
    created_by: str | None = None
    created_by_name: str | None = None
    created_at: datetime
    modified_at: datetime | None = None
    lines: list[InvoiceLineResponse] = []

    class Config:
        from_attributes = True


class PaginatedInvoices(BaseModel):
    items: list[InvoiceResponse]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# Payment Applications
# ---------------------------------------------------------------------------


class PaymentApplicationCreate(BaseModel):
    invoice_id: str
    amount_applied: Decimal


class PaymentApplicationResponse(BaseModel):
    id: str
    payment_id: str
    invoice_id: str
    invoice_number: str | None = None
    amount_applied: Decimal

    class Config:
        from_attributes = True


# ---------------------------------------------------------------------------
# Customer Payments
# ---------------------------------------------------------------------------


class CustomerPaymentCreate(BaseModel):
    customer_id: str
    payment_date: datetime
    total_amount: Decimal
    payment_method: str
    reference_number: str | None = None
    notes: str | None = None
    applications: list[PaymentApplicationCreate] = []


class CustomerPaymentResponse(BaseModel):
    id: str
    company_id: str
    customer_id: str
    customer_name: str | None = None
    payment_date: datetime
    total_amount: Decimal
    payment_method: str
    reference_number: str | None = None
    notes: str | None = None
    created_by: str | None = None
    created_by_name: str | None = None
    created_at: datetime
    modified_at: datetime | None = None
    applications: list[PaymentApplicationResponse] = []

    class Config:
        from_attributes = True


class PaginatedCustomerPayments(BaseModel):
    items: list[CustomerPaymentResponse]
    total: int
    page: int
    per_page: int


# ---------------------------------------------------------------------------
# AR Aging Report
# ---------------------------------------------------------------------------


class ARAgingBucket(BaseModel):
    current: Decimal = Decimal("0.00")
    days_1_30: Decimal = Decimal("0.00")
    days_31_60: Decimal = Decimal("0.00")
    days_61_90: Decimal = Decimal("0.00")
    days_over_90: Decimal = Decimal("0.00")
    total: Decimal = Decimal("0.00")


class ARAgingCustomer(BaseModel):
    customer_id: str
    customer_name: str
    account_number: str | None = None
    buckets: ARAgingBucket


class ARAgingReport(BaseModel):
    company_summary: ARAgingBucket
    customers: list[ARAgingCustomer]


# ---------------------------------------------------------------------------
# Sales Stats
# ---------------------------------------------------------------------------


class SalesStats(BaseModel):
    total_quotes: int = 0
    open_quotes: int = 0
    total_orders: int = 0
    open_orders: int = 0
    total_invoices: int = 0
    outstanding_invoices: int = 0
    total_ar_outstanding: Decimal = Decimal("0.00")


# ---------------------------------------------------------------------------
# Payment Import
# ---------------------------------------------------------------------------


class PaymentImportResultRow(BaseModel):
    row: int
    message: str


class PaymentImportResult(BaseModel):
    created: int
    skipped: int
    errors: list[PaymentImportResultRow]

File: frontend/src/types/sales.ts
// ---------------------------------------------------------------------------
// Quote Lines
// ---------------------------------------------------------------------------

export interface QuoteLine {
  id: string;
  quote_id: string;
  product_id: string | null;
  product_name: string | null;
  description: string;
  quantity: string;
  unit_price: string;
  line_total: string;
  sort_order: number;
}

export interface QuoteLineCreate {
  product_id?: string | null;
  description: string;
  quantity: string;
  unit_price: string;
  sort_order?: number;
}

// ---------------------------------------------------------------------------
// Quotes
// ---------------------------------------------------------------------------

export interface Quote {
  id: string;
  company_id: string;
  number: string;
  customer_id: string;
  customer_name: string | null;
  status: string;
  quote_date: string;
  expiry_date: string;
  payment_terms: string | null;
  subtotal: string;
  tax_rate: string;
  tax_amount: string;
  total: string;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  modified_at: string | null;
  lines: QuoteLine[];
}

export interface QuoteCreate {
  customer_id: string;
  quote_date: string;
  expiry_date: string;
  payment_terms?: string | null;
  tax_rate?: string;
  notes?: string | null;
  lines: QuoteLineCreate[];
}

export interface QuoteUpdate {
  status?: string;
  expiry_date?: string;
  payment_terms?: string | null;
  notes?: string | null;
}

export interface PaginatedQuotes {
  items: Quote[];
  total: number;
  page: number;
  per_page: number;
}

// ---------------------------------------------------------------------------
// Sales Order Lines
// ---------------------------------------------------------------------------

export interface PersonalizationOption {
  type: "nameplate" | "cover_emblem" | "lifes_reflections" | "legacy_standard" | "legacy_custom";
  inscription_name?: string;
  inscription_dates?: string;
  inscription_additional?: string;
  print_name?: string;
  print_image_url?: string;
  symbol?: string;
  custom_description?: string;
}

export interface PersonalizationTask {
  id: string;
  task_type: string;
  inscription_name: string | null;
  inscription_dates: string | null;
  inscription_additional: string | null;
  print_name: string | null;
  print_image_url: string | null;
  symbol: string | null;
  is_custom_legacy: boolean;
  status: string;
  completed_at: string | null;
  notes: string | null;
}

export interface SalesOrderLine {
  id: string;
  sales_order_id: string;
  product_id: string | null;
  product_name: string | null;
  description: string;
  quantity: string;
  quantity_shipped: string;
  unit_price: string;
  line_total: string;
  sort_order: number;
  has_conditional_pricing?: boolean;
  is_call_office?: boolean;
  price_without_our_product?: string | null;
  personalization_data?: PersonalizationOption[] | null;
}

export interface SalesOrderLineCreate {
  product_id?: string | null;
  description: string;
  quantity: string;
  unit_price: string;
  sort_order?: number;
}

// ---------------------------------------------------------------------------
// Sales Orders
// ---------------------------------------------------------------------------

export interface SalesOrder {
  id: string;
  company_id: string;
  number: string;
  customer_id: string;
  customer_name: string | null;
  quote_id: string | null;
  status: string;
  order_date: string;
  required_date: string | null;
  shipped_date: string | null;
  payment_terms: string | null;
  subtotal: string;
  tax_rate: string;
  tax_amount: string;
  total: string;
  ship_to_name: string | null;
  ship_to_address: string | null;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  modified_at: string | null;
  lines: SalesOrderLine[];
  driver_confirmed?: boolean;
  completed_at?: string | null;
  has_inventory_warning?: boolean;
  inventory_warning_notes?: string | null;
  has_driver_exception?: boolean;
  delivered_at?: string | null;
  service_location?: string | null;
  service_location_other?: string | null;
  service_time?: string | null;
  eta?: string | null;
}

export interface SalesOrderCreate {
  customer_id: string;
  quote_id?: string | null;
  order_date: string;
  required_date?: string | null;
  payment_terms?: string | null;
  tax_rate?: string;
  ship_to_name?: string | null;
  ship_to_address?: string | null;
  notes?: string | null;
  lines: SalesOrderLineCreate[];
}

export interface SalesOrderUpdate {
  status?: string;
  required_date?: string | null;
  shipped_date?: string | null;
  notes?: string | null;
}

export interface PaginatedSalesOrders {
  items: SalesOrder[];
  total: number;
  page: number;
  per_page: number;
}

// ---------------------------------------------------------------------------
// Invoice Lines
// ---------------------------------------------------------------------------

export interface InvoiceLine {
  id: string;
  invoice_id: string;
  product_id: string | null;
  product_name: string | null;
  description: string;
  quantity: string;
  unit_price: string;
  line_total: string;
  sort_order: number;
}

export interface InvoiceLineCreate {
  product_id?: string | null;
  description: string;
  quantity: string;
  unit_price: string;
  sort_order?: number;
}

// ---------------------------------------------------------------------------
// Invoices
// ---------------------------------------------------------------------------

export interface Invoice {
  id: string;
  company_id: string;
  number: string;
  customer_id: string;
  customer_name: string | null;
  sales_order_id: string | null;
  status: string;
  invoice_date: string;
  due_date: string;
  payment_terms: string | null;
  subtotal: string;
  tax_rate: string;
  tax_amount: string;
  total: string;
  amount_paid: string;
  balance_remaining: string;
  notes: string | null;
  deceased_name: string | null;
  sent_at: string | null;
  sent_to_email: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  modified_at: string | null;
  lines: InvoiceLine[];
}

export interface InvoiceCreate {
  customer_id: string;
  sales_order_id?: string | null;
  invoice_date: string;
  due_date: string;
  payment_terms?: string | null;
  tax_rate?: string;
  notes?: string | null;
  lines: InvoiceLineCreate[];
}

export interface InvoiceUpdate {
  status?: string;
  notes?: string | null;
}

export interface PaginatedInvoices {
  items: Invoice[];
  total: number;
  page: number;
  per_page: number;
}

// ---------------------------------------------------------------------------
// Payment Applications
// ---------------------------------------------------------------------------

export interface PaymentApplication {
  id: string;
  payment_id: string;
  invoice_id: string;
  invoice_number: string | null;
  amount_applied: string;
}

export interface PaymentApplicationCreate {
  invoice_id: string;
  amount_applied: string;
}

// ---------------------------------------------------------------------------
// Customer Payments
// ---------------------------------------------------------------------------

export interface CustomerPayment {
  id: string;
  company_id: string;
  customer_id: string;
  customer_name: string | null;
  payment_date: string;
  total_amount: string;
  payment_method: string;
  reference_number: string | null;
  notes: string | null;
  created_by: string | null;
  created_by_name: string | null;
  created_at: string;
  modified_at: string | null;
  applications: PaymentApplication[];
}

export interface CustomerPaymentCreate {
  customer_id: string;
  payment_date: string;
  total_amount: string;
  payment_method: string;
  reference_number?: string | null;
  notes?: string | null;
  applications: PaymentApplicationCreate[];
}

export interface PaginatedCustomerPayments {
  items: CustomerPayment[];
  total: number;
  page: number;
  per_page: number;
}

// ---------------------------------------------------------------------------
// AR Aging
// ---------------------------------------------------------------------------

export interface ARAgingBucket {
  current: string;
  days_1_30: string;
  days_31_60: string;
  days_61_90: string;
  days_over_90: string;
  total: string;
}

export interface ARAgingCustomer {
  customer_id: string;
  customer_name: string;
  account_number: string | null;
  buckets: ARAgingBucket;
}

export interface ARAgingReport {
  company_summary: ARAgingBucket;
  customers: ARAgingCustomer[];
}

// ---------------------------------------------------------------------------
// Sales Stats
// ---------------------------------------------------------------------------

export interface SalesStats {
  total_quotes: number;
  open_quotes: number;
  total_orders: number;
  open_orders: number;
  total_invoices: number;
  outstanding_invoices: number;
  total_ar_outstanding: string;
}

// ---------------------------------------------------------------------------
// Payment Import
// ---------------------------------------------------------------------------

export interface PaymentImportResultRow {
  row: number;
  message: string;
}

export interface PaymentImportResult {
  created: number;
  skipped: number;
  errors: PaymentImportResultRow[];
}
